### PR TITLE
Homework 2 - PARK (REST API with Flask + Docker) 🐋 

### PR DESCRIPTION
**Pull request by sudotouchwoman**
Currently the main backend elements are implemented, namely:

- [x] flask application wraps online inference
- [x] app loads model artifact + auxiliary validation data on startup
- [x] app has `/health` endpoint to check whether the startup was done correctly (i.e., all the data was accessed and loaded)
- [x] `/predict` endpoint accepts requests with `payload` parameter containing `JSON` string (see `data/payload.json` for format example)
- [x] separate CLI application with simple web-client to request backend for prediction (see `client.py`)
- [x] updated project structure at top-level `README` and `docs/` directory with usage examples
- [x] simple validation steps: check the table format, numeric/categorical feature presense, column order and simple 6-sigma distribution test for numeric features
- [x] wrap server into `Docker` and compose
- [x] unit-tests for flask (these seem cumbersome and overcomplicated but must be a good expierence (like, setup/teardown components, `yield` fixtures...)
- [x] collect data from urls on server startup (currently commited links lead to my Google Drive)
- [x] more flexibility for external server configuration with envars like logging level and handlers (`stderr`, file...)
- [x] further image size optimization
(shrinked from `~650Mb` down to `~515Mb` by copying required sources only from host and `--no-cache-dir` flag for `pip`)

Frankly speaking, I tried to do my best in terms of design and robustness, but there still are some things to do (would be great to do) like fetching data from `s3` storage, but I do not have much time to work around this 😞 

Waiting for the feedback!

### CODE REVIEW

**Comment by sudotouchwoman** [comment]
The PR was closed automatically as `homework1` branch is no longer alive. Will be reopened into `main` branch

**Comment by SVizor42** [comment]
I see you made some optimizations with the docker image. Guess you could mark it as done. Please, add self-evaluation part (each point and overall sum), just to avoid miscommunications.

File: online_inference/app/__init__.py [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+
+
+def default_logger(name: str) -> logging.Logger:
+    return make_logger(name, LOGFILE)
+
+
+log = default_logger(__name__)
+
+
+def make_app(s: AppConfig = AppConfig()) -> Flask:
```
Comment (SVizor42) [inline online_inference/app/__init__.py]: `s` = `settings`?

File: online_inference/docs/README.md [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+
+Application can be accessed from the host via testing client too. The only thing to consider is that
+client will of course try to access the app at the host port, not the container's one.
+
+__UPD: pull the image from Docker Hub__
+
+Image was pushed to Docker hub, information is avaliable [there](https://hub.docker.com/r/sudotouchwoman/wbcd-online-inference).
+
+```
+$ docker pull sudotouchwoman/wbcd-online-inference
```
Comment (SVizor42) [inline online_inference/docs/README.md]: Note: docker image with the `latest` TAG isn't available on Dockerhub

Comment (sudotouchwoman) [reply online_inference/docs/README.md]: Guess I should hardcode this command to pull `v1` or `v2` then
I've read several articles on this topic and sometimes usage of `latest` TAG is discouraged unless one really maintains the image

File: online_inference/app/__init__.py (hunk 1 above)
Comment (sudotouchwoman) [reply online_inference/app/__init__.py]: Yes, will be renamed to `settings` to make more sense

**Comment by sudotouchwoman** [comment]
Regarding the self-evaluation (will dublicate some points here)

__Base part__
+ Wrap online infrerence with a restful Flask service, `/predict` endpoint present `(3/3)`
+ `/health` endpoint to test whether the model is ready to operate `(1/1)` _Note:_ In this implementation, I did not configure the app to respond with 400, actually. It would rather respond with `{ "status": 400 }`. I modified this behavior while working with `k8s` (although now I suppose it would be better to have an endpoint like `/alive` or smth which would always respond to perform liveness probes)
+ Unit-tests for `/predict` `(3/3)`: currently testing startup and all present endpoints
+ Script to work  with the API `(2/2)`: created a separate dummy CLI client (I was thinking of `streamlit`-based web client but had no idea of its functionality as the features in my demo make no sense)
+ `Dockerfile` for the API with usage examples `(4/4)`
+ Push the image to hub `(2/2)`
+ Provide commands to pull image from hub `(1/1)`
+ Self-evaluation

__Additional activities__
+ The service fetches model from `s3`/other external storage, path to the resource is passed via envars `(1/2)`: I do not load the model specifically from `s3` using credentials, rather by a url (application can load resources either from the container's local filesystem or from url. It is actually possible to mount the data and provide the mount path via envars too)
+ Docker image size optimization `(1/2)`: I use `--no-cache-dir` to drop `pip` cache, which managed to take about `100Mb`.
+ Input validation `(2/2)`: I validate the input format, specifically the column data types followed by a dummy 6-sigma outlier check. When the column format validation is failed, I respond with `"prediction": null`. Statistics and column format are loaded in the same manner as the model artifact. Also, I validate that the artifact loaded is an instance of `sklearn.Pipeline` and has the `.predict` method.

Overall sum: `21/23`

**Comment by SVizor42** [comment]
Well done! 👍 

Just a couple of comments:
1. Please, also add to README file information about setting up `ENV_PATH` before running `flask run` from the terminal (in case I want to use predefined paths).
2. We can grade docker optimization part as 2/2 - you also took small base image and tried to minimize layers number.
3. Actually, I'm not sure that everyone is going to download model from s3-like storage (with authorization). The model is successfully loading from the gdrive using the provided links from predefined env list. I suggest to grade it as 1.5 of 2.

**Total: 22.5 points**

Merge please, if you have no objections.

**Comment by sudotouchwoman** [comment]
Added `ENV_PATH` usage example to `README`, proceeding to merge

**Pull request merged**